src/convert.civet

```diff
@@ -0,0 +1,553 @@
+  fills in the corresponding `faces_edges` property (preserving order).
+  ###
+  edgeMap: Record<string, number> := {}
+  for [v1, v2], edge of fold.edges_vertices
+    edgeMap[`${v1},${v2}`] = edge
+    edgeMap[`${v2},${v1}`] = edge
+  fold.faces_edges =
+    for vertices, face of fold.faces_vertices
+      for i of [0...vertices.length]
+        edgeMap[`${vertices[i]},${vertices[(i+1) % vertices.length]}`]
+
+export function faces_vertices_to_edges(mesh: Fold)
+  ###
+  Given a FOLD object with just `faces_vertices`, automatically fills in
+  `edges_vertices`, `edges_faces`, `faces_edges`, and `edges_assignment`
+  (indicating which edges are boundary with 'B').
+  This code currently assumes an orientable manifold, and uses nulls to
+  represent missing neighbor faces in `edges_faces` (for boundary edges).
+  ###
+  mesh.edges_vertices = []
+  mesh.edges_faces = []
+  mesh.faces_edges = []
+  mesh.edges_assignment = []
+
+  edgeMap: Record<string, number> := {}
+
+  for face_key, vertices in mesh.faces_vertices
+    face := parseInt face_key
+    mesh.faces_edges.push(
+      for v1_key, i of vertices
+        v1 := parseInt v1_key
+        v2 := vertices[(i+1) % vertices.length]
+
+        key: string .= ""
+        if v1 <= v2
+          key = `${v1},${v2}`
+        else
+          key = `${v2},${v1}`
+
+        edge: number .= null
+        if key in edgeMap
+          edge = edgeMap[key]
+          // Second instance of edge means not on boundary
+          mesh.edges_assignment[edge] = null
+        else
+          edge = edgeMap[key] = mesh.edges_vertices.length
+          if v1 <= v2
+            mesh.edges_vertices.push [v1, v2]
+          else
+            mesh.edges_vertices.push [v2, v1]
+          mesh.edges_faces.push [null, null]
+          // First instance of edge might be on boundary
+          mesh.edges_assignment.push 'B'
+
+        if v1 <= v2
+          mesh.edges_faces[edge][0] = face
+        else
+          mesh.edges_faces[edge][1] = face
+        edge
+    )
+  mesh
+
+export function edges_vertices_to_edges_faces_edges(fold: Fold)
+  ###
+  Given a `fold` object with `edges_vertices` and `faces_vertices`,
+  fills in `faces_edges` and `edges_vertices`.
+  ###
+  fold.edges_faces = ([null, null] for edge of [0...fold.edges_vertices.length])
+  edgeMap: Record<string, Coords> := {}
+
+  for edge_key, vertices in fold.edges_vertices
+    continue unless vertices?
+    edge := parseInt edge_key
+    edgeMap[`${vertices[0]},${vertices[1]}`] = [edge, 0] // forward
+    edgeMap[`${vertices[1]},${vertices[0]}`] = [edge, 1] // backward
+
+  fold.faces_edges ?= []
+  for each vertices, face of fold.faces_vertices
+    fold.faces_edges[face] =
+      for v1, i of vertices
+        v2 := vertices[(i+1) % vertices.length]
+        [edge, orient] := edgeMap[`${v1},${v2}`]
+        fold.edges_faces[edge][orient] = face
+        edge
+  fold
+
+export function unfoldedGeometry(fold: Fold, rootFace = 0)
```

Comment:
There is a ton of shared code between these three functions. Can we factor them out into a generic function that gets called three different ways?